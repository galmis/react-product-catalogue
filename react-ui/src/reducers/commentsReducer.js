// @flow

import {
  FETCH_COMMENTS_SUCCESS,
  CREATE_COMMENT_SUCCESS,
  REPLY_COMMENT
} from '../constants/ACTION_TYPE';
import type { Action, NormalizedData } from '../types';

const initialState: Object = {
  commentsById: {},
  totalComments: 0,
  totalPages: 0,
  fetchedThreads: {},
  createdThreads: {},
  fetchedPosts: {},
  commentToReplyId: 0,
  postId: ''
};

export {
  initialState
}

export default function commentsReducer(state: Object = initialState, action: Action) {

  switch(action.type) {
    case FETCH_COMMENTS_SUCCESS: {

      const {postId, parentId, data} = action.payload;
      const fetchedThreads = { ...state.fetchedThreads };
      const totalComments = parentId >= 0 ? state.totalComments : action.payload.totalRecords;
      const totalPages = parentId >= 0 ? state.totalPages : action.payload.totalPages;

      if (parentId >= 0) {
        const threadId = parentId === 0 ? '0:' + postId : parentId.toString();
        const threadData = { ...fetchedThreads[threadId] };
        const currReplies = threadData.replies ? threadData.replies : [];
        const result = data ? data.result.filter(id => currReplies.indexOf(id) === -1) : [];

        fetchedThreads[threadId] = {
          replies: [...currReplies, ...result],
          totalReplies: action.payload.totalRecords,
        };
      }
      let commentsById = state.commentsById;
      if (data) {
        commentsById = {
          ...commentsById,
          ...action.payload.data.entities.dataById
        };
      }

      const newState =  {
        ...state,
        postId,
        fetchedThreads,
        totalComments,
        totalPages,
        commentsById
      };

      return newState;
    }
    case REPLY_COMMENT: {
      return {
        ...state,
        commentToReplyId: action.payload.commentToReplyId
      }
    }
    case CREATE_COMMENT_SUCCESS: {
      const {postId, parentId, data} = action.payload;

      const result = data.result;
      const threadId = parentId === 0 ? '0:' + postId : parentId.toString();
      const createdThreads = { ...state.createdThreads };
      const threadData = createdThreads[threadId];
      const currReplies = threadData ? threadData.replies : [];
      const replies = parentId === 0 ? [...result, ...currReplies]
        : [...currReplies, ...result];

      createdThreads[threadId] = {
        replies
      }

      const commentsById = {
        ...state.commentsById,
        ...data.entities.dataById
      };

      return {
        ...state,
        totalComments: state.totalComments + 1,
        commentsById,
        createdThreads
      }
    }
  }

  return state;
}
