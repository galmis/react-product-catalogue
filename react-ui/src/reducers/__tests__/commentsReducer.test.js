// @flow
'use strict';

jest.dontMock('../commentsReducer.js');

import commentsReducer from '../commentsReducer';
import deepFreeze from 'deep-freeze';

import type { Action } from '../../types';

import {
  FETCH_COMMENTS_SUCCESS,
  REPLY_COMMENT,
  CREATE_COMMENT_SUCCESS
} from '../../constants/ACTION_TYPE';

describe('commentsReducer', () => {

  it('FETCH_COMMENTS_SUCCESS, initialState', () => {
    const action: Action = {
      type: FETCH_COMMENTS_SUCCESS,
      payload: {
        data: {
          result: ['1'],
          entities: {
            dataById: {
              '1': {
                parent: 0,
                id: 1,
                post: 2,
                content: {
                  rendered: '<p>test</p>'
                }
              }
            }
          },
        },
        totalRecords: 1,
        totalPages: 1,
        parentId: 0,
        postId: '2'
      }
    };
    deepFreeze(action);

    expect(commentsReducer(undefined, action)).toMatchSnapshot();
  });


  it('FETCH_COMMENTS_SUCCESS, rewrite existing comment', () => {
    const action: Action = {
      type: FETCH_COMMENTS_SUCCESS,
      payload: {
        data: {
          result: ['1'],
          entities: {
            dataById: {
              '1': {
                parent: 0,
                id: 1,
                post: 2,
                content: {
                  rendered: '<p>new</p>'
                }
              }
            }
          },
        },
        totalRecords: 1,
        totalPages: 1,
        parentId: 0,
        postId: '2'
      }
    };
    deepFreeze(action);


    const oldState = {
      commentsById: {
        '1': {
          parent: 0,
          id: 1,
          post: 2,
          content: {
            rendered: '<p>existing</p>'
          }
        }
      },
      totalComments: 1,
      totalPages: 1,
      fetchedThreads: {
        '0:2': {
          replies: ['1'],
          totalReplies: 1
        }
      },
      createdThreads: {},
      fetchedPosts: {},
      commentToReplyId: 0,
      postId: '2'
    }
    deepFreeze(oldState);

    expect(commentsReducer(oldState, action)).toMatchSnapshot();
  });

  it('REPLY_COMMENT', () => {
    const action: Action = {
      type: REPLY_COMMENT,
      payload: {
        commentToReplyId: 1
      }
    };
    deepFreeze(action);

    const oldState = {
      commentsById: {
        '1': {
          parent: 0,
          id: 1,
          post: 2,
          content: {
            rendered: '<p>existing</p>'
          }
        }
      },
      totalComments: 1,
      totalPages: 1,
      fetchedThreads: {
        '0:2': {
          replies: ['1'],
          totalReplies: 1
        }
      },
      createdThreads: {},
      fetchedPosts: {},
      commentToReplyId: 0,
      postId: '2'
    }
    deepFreeze(oldState);

    expect(commentsReducer(oldState, action)).toMatchSnapshot();
  });

  it('CREATE_COMMENT_SUCCESS', () => {
    const action: Action = {
      type: CREATE_COMMENT_SUCCESS,
      payload: {
        data: {
          result: ['2'],
          entities: {
            dataById: {
              '2': {
                parent: 0,
                id: 2,
                post: 2,
                content: {
                  rendered: '<p>new</p>'
                }
              }
            }
          },
        },
        parentId: 0,
        postId: '2',
        status: 201
      }
    };
    deepFreeze(action);

    const oldState = {
      commentsById: {
        '1': {
          parent: 0,
          id: 1,
          post: 2,
          content: {
            rendered: '<p>existing</p>'
          }
        }
      },
      totalComments: 1,
      totalPages: 1,
      fetchedThreads: {
        '0:2': {
          replies: ['1'],
          totalReplies: 1
        }
      },
      createdThreads: {},
      fetchedPosts: {},
      commentToReplyId: 0,
      postId: '2'
    }
    deepFreeze(oldState);

    expect(commentsReducer(oldState, action)).toMatchSnapshot();
  });

});
