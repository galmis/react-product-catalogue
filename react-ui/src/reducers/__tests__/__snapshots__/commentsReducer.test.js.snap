exports[`commentsReducer CREATE_COMMENT_SUCCESS 1`] = `
Object {
  "commentToReplyId": 0,
  "commentsById": Object {
    "1": Object {
      "content": Object {
        "rendered": "<p>existing</p>",
      },
      "id": 1,
      "parent": 0,
      "post": 2,
    },
    "2": Object {
      "content": Object {
        "rendered": "<p>new</p>",
      },
      "id": 2,
      "parent": 0,
      "post": 2,
    },
  },
  "createdThreads": Object {
    "0:2": Object {
      "replies": Array [
        "2",
      ],
    },
  },
  "fetchedPosts": Object {},
  "fetchedThreads": Object {
    "0:2": Object {
      "replies": Array [
        "1",
      ],
      "totalReplies": 1,
    },
  },
  "postId": "2",
  "totalComments": 2,
  "totalPages": 1,
}
`;

exports[`commentsReducer FETCH_COMMENTS_SUCCESS, initialState 1`] = `
Object {
  "commentToReplyId": 0,
  "commentsById": Object {
    "1": Object {
      "content": Object {
        "rendered": "<p>test</p>",
      },
      "id": 1,
      "parent": 0,
      "post": 2,
    },
  },
  "createdThreads": Object {},
  "fetchedPosts": Object {},
  "fetchedThreads": Object {
    "0:2": Object {
      "replies": Array [
        "1",
      ],
      "totalReplies": 1,
    },
  },
  "postId": "2",
  "totalComments": 0,
  "totalPages": 0,
}
`;

exports[`commentsReducer FETCH_COMMENTS_SUCCESS, rewrite existing comment 1`] = `
Object {
  "commentToReplyId": 0,
  "commentsById": Object {
    "1": Object {
      "content": Object {
        "rendered": "<p>new</p>",
      },
      "id": 1,
      "parent": 0,
      "post": 2,
    },
  },
  "createdThreads": Object {},
  "fetchedPosts": Object {},
  "fetchedThreads": Object {
    "0:2": Object {
      "replies": Array [
        "1",
      ],
      "totalReplies": 1,
    },
  },
  "postId": "2",
  "totalComments": 1,
  "totalPages": 1,
}
`;

exports[`commentsReducer REPLY_COMMENT 1`] = `
Object {
  "commentToReplyId": 1,
  "commentsById": Object {
    "1": Object {
      "content": Object {
        "rendered": "<p>existing</p>",
      },
      "id": 1,
      "parent": 0,
      "post": 2,
    },
  },
  "createdThreads": Object {},
  "fetchedPosts": Object {},
  "fetchedThreads": Object {
    "0:2": Object {
      "replies": Array [
        "1",
      ],
      "totalReplies": 1,
    },
  },
  "postId": "2",
  "totalComments": 1,
  "totalPages": 1,
}
`;
